### PR TITLE
add GPU tests on Bowie cluster

### PR DESCRIPTION
**Pull request by Delcior**

### CODE REVIEW

File: .github/workflows/bowie.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          repository: ${{ github.event.pull_request.head.repo.full_name || github.repository }}
+          ref: ${{ github.event.pull_request.head.sha || github.sha }}
+          allow-unsafe-pr-checkout: true
+          persist-credentials: false
+
+      - name: Run GPU Tests
+        uses: ./.github/actions/run-slurm-job
+        with:
+          python-version: 3.13.13 # DO NOT CHANGE
+          command: "env -u CI NUMBA_THREADING_LAYER=workqueue python -m pytest --durations=10 -v -s -p no:unraisableexception --timeout=900 --timeout_method=thread ${{ inputs.tests_to_run || 'tests/unit_tests' }} -k 'GPU or ThrustRTC'"
```
Comment (slayoo) [inline .github/workflows/bowie.yml:61]: the new JAX backend also support GPU
anyhow, wouldn't it be simpler to just run all tests?

File: .github/workflows/bowie.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        with:
+          repository: ${{ github.event.pull_request.head.repo.full_name || github.repository }}
+          ref: ${{ github.event.pull_request.head.sha || github.sha }}
+          allow-unsafe-pr-checkout: true
+          persist-credentials: false
+
+      - name: Run GPU Tests
+        uses: ./.github/actions/run-slurm-job
+        with:
+          python-version: 3.13.13 # DO NOT CHANGE
```
Comment (slayoo) [inline .github/workflows/bowie.yml:60]: worth explaining why not to change

File: .github/scripts/run-bowie-tests.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#SBATCH --gres=gpu:1
+
+source $HOME/.setup-pyenv
+pyenv shell $1
+python --version
+python -m venv ./.venv
+source ./.venv/bin/activate
+
+pip install uv
+uv pip install numpy==1.26.4
```
Comment (slayoo) [inline .github/scripts/run-bowie-tests.sh:21]: JAX requires NumPy>2
is there a particular reason to use 1.26.4 here?

File: .github/workflows/bowie.yml (hunk 1 above)
Comment (Delcior) [reply .github/workflows/bowie.yml:61]: There's a lot of failing tests and output on UI is absolutely unreadable. I thought it'd be nice to have an option to run particular subset of tests or even a signle test to make it easier to debug (this feature works only when you trigger GPU Tests via `workflow_dispatch`. For PRs it always runs all the tests). 

Two questions:
1. Do you want to keep 'run custom tests' feature for `workflow_dispatch`?
2. By "just run all the tests" you mean **tests/unit_tests** or **tests/**?

File: .github/workflows/bowie.yml (hunk 2 above)
Comment (Delcior) [reply .github/workflows/bowie.yml:60]: I'll add a note that cluster has support only for python 3.13.13

File: .github/scripts/run-bowie-tests.sh (hunk 3 above)
Comment (Delcior) [reply .github/scripts/run-bowie-tests.sh:21]: I was working on some old version of PySDM that had pinned numpy to 1.26.4 (for CI builds). When I was playing with unsetting `CI` variable to run GPU tests I was sometimes getting higher version of numpy and it had breaking interactions with other packages so I pinned it in the script purely for the dev process.

I've just triggered GPU tests after rebasing to main and with this line deleted and it worked, so I'll just delete it for good. Good catch!

File: .github/workflows/bowie.yml (hunk 1 above)
Comment (slayoo) [reply .github/workflows/bowie.yml:61]: I meant the `-k` option

Comment (Delcior) [reply .github/workflows/bowie.yml:61]: With `-k 'GPU or ThrustRTC'` we have:
- pipeline duration: ~7-8 minutes
- tests collected: ~400

**without** `-k 'GPU or ThrustRTC'` we have:
- pipeline duration: **> 1h** (I don't know exactly how long, because I hit my timeout limit set in workflow. [Link](https://github.com/Delcior/PySDM/actions/runs/33677798379))
- tests collected: **~2000**
- duplicate work (as these tests already run in other workflows)

I see 3 options:
1. Leave `-k GPU or ThrustRTC'`. AI told me that this option currently omits 4 GPU tests.
2. Leave `-k GPU or ThrustRTC'`, but start an effort to mark GPU tests using [pytest markers](https://docs.pytest.org/en/stable/example/markers.html#marking-test-functions-and-selecting-them-for-a-run) so that we can select the GPU tests via `-m` option (much more reliable than selecting by name as we currently attempt to do)
3. Increase the timeout, remove `-k GPU or ThrustRTC'` and run all the tests with one run taking > 1h. 

I have very limited knowledge about PySDM codebase, but I think the safest and most profitable option is 2.
What do you think?

Side note: `pytest` runs the tests on a single core, so I installed [pytest-xdist](https://pypi.org/project/pytest-xdist/) to use all 4 cores from the Jetson node, but it didn't help much as it also hit the 1h timeout [link](https://github.com/Delcior/PySDM/actions/runs/33684452832). (not saying that other options/optimizations wouldn't help, just highlighting our hardware limitations).

Comment (slayoo) [reply .github/workflows/bowie.yml:61]: Regarding multi-threading, I'd expect that pytest parallelisation will not help much as most of the PySDM logic uses multi-threading internally.

I agree that there is no point in running all tests on Bowie, since we can test on ARM64 on GitHub. The point is how to pick GPU tests. These are effectively already marked with GPU via the `backend_class` and `backend_instance` fixtures which resolve to `GPU`, hence "-k GPU" is kind of OK. However, the direction we're embracing is to phase out the ThrustRTC backend at all, in favour of the new JAX backend (already in `main`), and the `-k GPU` skips JAX (even if we flag with `GPU` label). So, for now, let's perhaps do -k `GPU or ThrustRTC or JAX`?

@Bodzio-2, how to make JAX tests actually use the GPU?

Comment (Delcior) [reply .github/workflows/bowie.yml:61]: > [...] start an effort to mark GPU tests using [pytest markers](https://docs.pytest.org/en/stable/example/markers.html#marking-test-functions-and-selecting-them-for-a-run) [...]

and what about that option? do you think it's worth the effort? Using `-k` right now is totally fine, but in longterm maybe it's better to lean towards `-m`?

Comment (slayoo) [reply .github/workflows/bowie.yml:61]: IIUC, all GPU tests we have are triggered via the backend fixtures, hence marking them in another way seems redundant

Comment (Bodzio-2) [reply .github/workflows/bowie.yml:61]: JAX has a config option to set the [default device](https://docs.jax.dev/en/latest/_autosummary/jax.default_device.html). That could be fit into the backend initialization to select based on user preference. Also, from what I've read JAX doesn't automatically do any multi-threading on CPU.

*GPU on JAX works only on Linux or dubiously via WSL

Comment (slayoo) [reply .github/workflows/bowie.yml:61]: relying on a "default" setting could be tricky as we do want to be able to run JAX-CPU and JAX-GPU simulations in one notebook, or at least warn the user if such setting is not possible

Comment (Bodzio-2) [reply .github/workflows/bowie.yml:61]: You can run code on a specific device wrapping it with a: ```with jax.default_device(jax.devices("gpu")[0]):```.

For a more user-friendly experience I'd imagine an argument to the JAX backend constructor to attempt using gpu/cpu (with a warning/exception if a device wasn't found).

Comment (slayoo) [reply .github/workflows/bowie.yml:61]: would something like this be OK: https://github.com/open-atmos/PySDM/pull/1941
(haven't tested, just a draft)

Comment (Delcior) [reply .github/workflows/bowie.yml:62]: **TLDR**: GPU tests with JAX **will not work at all** on our current Bowie Cluster due to CUDA == 10.2.

PySDM uses [JAX](https://github.com/open-atmos/PySDM/blob/main/pyproject.toml#L44) >= 0.8.0, but according to the [documentation](https://docs.jax.dev/en/latest/installation.html) this package alone gives support for CPU-only backend. We'd need to install `jax[cuda*]` to get the GPU one. 

JAX supports CUDA >= 12 ([docs](https://docs.jax.dev/en/latest/installation.html#install-nvidia-gpu)). Support for CUDA 10.x was dropped ~5 years ago ([link](https://docs.jax.dev/en/latest/changelog.html#jaxlib-0-1-72-oct-12-2021)) and our worker nodes are equipped with CUDA==10.2 which we **cannot upgrade**. That means, in the current setup, we will not be able to run these JAX GPU tests on the Bowie Cluster.
